File: launcher/src/main/java/com/ethan/launcher/launch/HolderFragment.java
package com.ethan.launcher.launch;

import android.content.Intent;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import com.ethan.launcher.MainLauncherActivity;

import static com.ethan.launcher.launch.LauncherConstance.LAUNCHER_CONFIG_SETTING;

public class HolderFragment extends Fragment {
    private static final String TAG = "com.launcher.launcher";

    private static final int LAUNCHER_START_REQUEST_CODE = 0x66;

    private LauncherCallBack mLauncherCallBack;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    public static HolderFragment getHolderFragment(FragmentActivity fragmentActivity) {
        return getHolderFragment(fragmentActivity.getSupportFragmentManager());
    }

    public static HolderFragment getHolderFragment(Fragment fragment) {
        return getHolderFragment(fragment.getChildFragmentManager());
    }

    private static HolderFragment getHolderFragment(FragmentManager fragmentManager) {
        HolderFragment holderFragment = findHolderFragment(fragmentManager);
        if (holderFragment == null) {
            holderFragment = new HolderFragment();
            fragmentManager.beginTransaction()
                    .add(holderFragment, TAG)
                    .commitAllowingStateLoss();
            fragmentManager.executePendingTransactions();
        }
        return holderFragment;
    }

    private static HolderFragment findHolderFragment(FragmentManager fragmentManager) {
        return (HolderFragment) fragmentManager.findFragmentByTag(TAG);
    }


    public void startLauncher(LauncherCallBack callBack, LauncherConfig launcherConfig) {
        this.mLauncherCallBack = callBack;
        Intent intent = new Intent(getContext(), MainLauncherActivity.class);
        intent.putExtra(LAUNCHER_CONFIG_SETTING, launcherConfig);
        startActivityForResult(intent, LAUNCHER_START_REQUEST_CODE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if ((requestCode == LAUNCHER_START_REQUEST_CODE) && (resultCode == 10086)) {
            if (mLauncherCallBack != null) {
                mLauncherCallBack.onLauncherResult(10088);
            }
        }
    }
}

File: launcherkt/src/main/java/com/ethan/launcherkt/launch/HolderFragment.kt
package com.ethan.launcherkt.launch

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import com.ethan.launcherkt.LauncherKtMainActivity

class HolderFragment : Fragment() {

    private lateinit var mLaunchCallBack: LaunchCallBack

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    fun launchActivity(callBack: LaunchCallBack) {
        this.mLaunchCallBack = callBack
        val intent = Intent(context, LauncherKtMainActivity::class.java)
        startActivityForResult(intent, REQUEST_CODE_FOR_LAUNCH)
    }

    companion object Instance {
        private val TAG = "HolderFragment"
        private val REQUEST_CODE_FOR_LAUNCH = 10088

        fun getHolderFragment(fragment: Fragment): HolderFragment {
            return getHolderFragment(fragment.childFragmentManager)
        }

        fun getHolderFragment(activity: FragmentActivity): HolderFragment {
            return getHolderFragment(activity.supportFragmentManager)
        }

        private fun getHolderFragment(fragmentManager: FragmentManager): HolderFragment {
            var holderFragment = findHolderFragment(fragmentManager)
            if (holderFragment == null) {
                holderFragment = HolderFragment()
                fragmentManager.beginTransaction()
                    .add(holderFragment, TAG)
                    .commitAllowingStateLoss() // 执行
                fragmentManager.executePendingTransactions()
            }
            return holderFragment
        }

        private fun findHolderFragment(fragmentManager: FragmentManager?): HolderFragment? {
            if (fragmentManager != null) {
                return fragmentManager.findFragmentByTag(TAG) as HolderFragment?
            }
            return null
        }
    }


}